pkg/sql/sqlinstance/instancestorage/instancereader_test.go

```diff
@@ -188,14 +197,21 @@ func TestReader(t *testing.T) {
 					t.Fatal(err)
 				}
 			}
-			// Verify all instances are returned by GetAllInstances.
+			// Verify all instances are returned by GetAllInstances and GetAllInstancesUsingTxn.
 			testutils.SucceedsSoon(t, func() error {
 				instances, err := reader.GetAllInstances(ctx)
 				if err != nil {
 					return err
 				}
 				sortInstances(instances)
-				return testOutputFn(instanceIDs, rpcAddresses, sqlAddresses, sessionIDs, localities, instances)
+				if err := testOutputFn(instanceIDs, rpcAddresses, sqlAddresses, sessionIDs, localities, instances); err != nil {
+					return err
+				}
+				instancesUsingTxn, err := getInstancesUsingTxnWrapper()
+				if err != nil {
+					return err
+				}
+				return testOutputFn(instanceIDs, rpcAddresses, sqlAddresses, sessionIDs, localities, instancesUsingTxn)
```

Comment:
how about a helper to remove some of this boilerplate and to make the test easier to read?
Firstly, let's make a structure to pass to `testOutputFn`, let's call it `expectations`. Then, let's make a helper


```go
type expectations struct { /* ... */ }
testOutputFn := func(exp expectations, actualInstances []sqlinstance.InstanceInfo) { /* ... */ }

expectationsFromOffset := func(offset int) expectations {
    return expectations{
     instanceIDs: instanceIDs[offset:],
     rpcAddresses: rpcAddresses[offset:],
     sqlAddresses: sqlAddresses[offset:],
     sessionIDs: sessionIDs[offset:],
     localities: localities[offset:]
    }
}
// ...
verifyInstancesWithGetter := func(t *testing.T, name string, exp expectations, getInstances func()([]sqlinstance.InstanceInfo, error)) (retErr error) {
    defer func() { errors.Wrap(retErr, name) }
    instances, err := getInstances()
    if err != nil {
        return err
    }
    sortInstances(instances)
    return testOutputFn(exp, instances)
}
verifyInstances := func(t *testing.T, exp expectations) {
    if err := verifyInstancesWithGetter(t, "reader", offset, func() ([]sqlinstance.InstanceInfo, error) {
    		return reader.GetAllInstances(ctx)
    }); err != nil {
        return err
    }
    return verifyInstancesWithGetter(t, "txn", offset, func() ([]sqlinstance.InstanceInfo, error) {
    		return getInstancesUsingTxn(t)
    })
}

Reply:
Nice